custom_components/rct_power/__init__.py

```diff
@@ -112,44 +122,32 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> Literal[
     ):
         raise ConfigEntryNotReady
 
-    remove_update_listener = entry.add_update_listener(async_reload_entry)
-
-    domain_data[entry.entry_id] = RctPowerContext(
-        update_coordinators={
+    entry.runtime_data = RctData(
+        {
             EntityUpdatePriority.FREQUENT: frequent_update_coordinator,
             EntityUpdatePriority.INFREQUENT: infrequent_update_coordinator,
             EntityUpdatePriority.STATIC: static_update_coordinator,
-        },
-        clean_up=remove_update_listener,
+        }
     )
 
+    entry.async_on_unload(entry.add_update_listener(async_reload_entry))
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
     return True
 
 
-async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+async def async_unload_entry(hass: HomeAssistant, entry: RctConfigEntry) -> bool:
     """Handle removal of an entry."""
-    if (context := RctPowerContext.get_from_domain_data(hass, entry)) is None:
-        return False
-
-    unloaded = all(
-        await asyncio.gather(
-            *[
-                hass.config_entries.async_forward_entry_unload(entry, platform)
-                for platform in PLATFORMS
-            ]
-        )
-    )
-
-    if unloaded:
-        context.clean_up()
-        context.remove_from_domain_data(hass, entry)
-
-    return unloaded
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if (
+        unload_ok
+        and len(entries := hass.config_entries.async_loaded_entries(DOMAIN)) == 1
+        and entries[0].entry_id == entry.entry_id
+    ):
+        hass.data.pop(DOMAIN)
```

Comment:
💭 If we stopped tracking the message flag as mentioned above, this becomes superfluous too.